src/Lecture2.hs

```diff
@@ -74,7 +78,15 @@ return the removed element.
 >>> removeAt 10 [1 .. 5]
 (Nothing,[1,2,3,4,5])
 -}
-removeAt = error "TODO"
+removeAt :: Int -> [a] -> (Maybe a, [a])
+removeAt i xs = removeAt' i xs []
+
+removeAt' :: Int -> [a] -> [a] -> (Maybe a, [a])
+removeAt' _ [] ys = (Nothing, reverse ys)
+removeAt' 0 (x:xs) ys = (Just x, reverse ys ++ xs)
+removeAt' i (x:xs) ys
+          | i > 0 = removeAt' (i-1) xs (x:ys)
+          | otherwise = (Nothing, (x:xs)) 
```

Comment:
This is a valid implementation 👍🏻 

It has a few inefficient places:

* You compare `index` with 0 on every recursive step where it's enough to compare only once
* Adding an element to the end and calling `reverse`

You can implement this function without these drawbacks in the following more efficient way:

```suggestion
removeAt :: Int -> [a] -> (Maybe a, [a])
removeAt n l
    | n < 0 = (Nothing, l)
    | otherwise = go n l
  where
    go _ [] = (Nothing, [])
    go 0 (x : xs) = (Just x, xs)
    go i (x : xs) =
        let (val, rest) = go (i - 1) xs
        in (val, x : rest)
```

Reply:
Thanks for your feedback. I like your suggestion for the recursive function.